test/test_jit.py

```diff
@@ -1384,5 +1384,119 @@ def to_int(x) -> (y):
+            third = 0 # TODO: python lexical scoping
+            while i < lim:
+                third = first + second
+                first = second
+                second = third
+                j = 0
+                while j < 10:
+                    somenum = somenum * 2
+                    j = j + 1
+                i = i + j
+                i = i + dontmutateme
+
+            st = second + third
+            fs = first + second
+
+        ''')
+        self.assertExpected(str(cu.get_graph('test_while')))
+
+    def test_script_if(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_if(a, b) -> (c):
+            d = 3
+            if a > 10:
+                a = 3 + d
+            else:
+                b = 3 + d
+                d = 4
+            c = a + b
+        ''')
+        self.assertExpected(str(cu.get_graph('test_if')))
+
+    def test_script_if_noelse(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_if_noelse(a, b) -> (c):
+            if a > 10:
+                a = 3 + b
+            c = a + b
+        ''')
+        self.assertExpected(str(cu.get_graph('test_if_noelse')))
+
+    def test_script_while_nonexistant_value(self):
+        with self.assertRaisesRegex(RuntimeError, "undefined value x"):
+            cu = torch.jit._jit_script_compile('''
+            def test_while(a, b) -> (c):
+                while a < 10:
+                    a = a + x
+                    b = b + 1
+                c = a + b
+            ''')
+
+    def test_script_while_nonexistant_cond_value(self):
+        with self.assertRaisesRegex(RuntimeError, "undefined value x"):
+            cu = torch.jit._jit_script_compile('''
+            def test_while(a, b) -> (c):
+                while a < x:
+                    a = a + 1
+                    b = b + 1
+                c = a + b
+            ''')
+
+    def test_script_while_write_outer_then_read(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_while(a, b) -> (c):
+            while a < 10:
+                a = a + 1
+                b = a + 1
+            c = a + b
+        ''')
+        self.assertExpected(str(cu.get_graph('test_while')))
+
+    def test_script_while_nest_if(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_while_if(a, b) -> (c):
+            c = 0
+            while a < 10:
+                a = a + 1
+                b = b + 1
+                if a > b:
+                    c = -a
+                else:
+                    c = -b
+            c = c + 1
+        ''')
+        self.assertExpected(str(cu.get_graph('test_while_if')))
+
+    def test_script_if_nest_while(self):
+        cu = torch.jit._jit_script_compile('''
+        def test_if_while(a, b) -> (c):
+            c = 0
+            if a > b:
+                while a > b:
+                    b = b + 1
+                    c = -b
+        ''')
+        self.assertExpected(str(cu.get_graph('test_if_while')))
```

Comment:
I’m ok with having one or two expects for this, but I’d rather add support for control-flow in the interpreter, and compare script to Python, rather than dump a bunch of large expect files. They don’t scale well, need to be periodically updated, and no one is really checking them carefully when this happens

Reply:
I know I am going to be closely reading the expect files when I do review. Perhaps my only complaint is that they are not inline with the corresponding tests... but that can be fixed with a little hacking :)

Reply:
@apaszke yeah I don't like having all these expects but this PR is meant to break our chicken-and-egg problem wrt control flow in the front-end vs interpreter. Expect these expects to disappear once we have interpreter support

Reply:
@ezyang that's good, but I remember at least a few situations when we were merging last patches that were changing tens of expect files, only to realize two weeks later that in fact some of them no longer look as they should. From what I've seen this is the least effective form of testing we have in our codebase.